src/JsonReader.g.cs

```diff
@@ -304,51 +305,42 @@ partial class JsonReader
             IJsonReader<T3> item3Reader) =>
         CreatePure((ref Utf8JsonReader rdr) =>
         {
-            if (rdr.TokenType is JsonTokenType.None && !rdr.Read())
-                throw PartialJsonNotSupportedException();
-
-            if (rdr.TokenType is not JsonTokenType.StartArray)
-                return Error("Invalid JSON value where a JSON array was expected.");
-
-            if (!rdr.Read())
-                throw PartialJsonNotSupportedException();
+            var (sm, item1, item2, item3) =
+                rdr.IsResuming ? ((ArrayReadStateMachine, T1, T2, T3))rdr.Pop() : default;
 
-            T1 item1;
-            switch(item1Reader.TryRead(ref rdr))
+            while (true)
             {
-                case { Incomplete: true }: throw PartialJsonNotSupportedException();
-                case { Error: { } error1 }: return Error(error1);
-                case { Value: var item }: item1 = item; break;
+                switch (sm.Read(ref rdr))
+                {
+                    case ArrayReadStateMachine.ReadResult.Error:
+                        return Error("Invalid JSON value where a JSON array was expected.");
+
+                    case ArrayReadStateMachine.ReadResult.Incomplete:
+                        return rdr.Suspend((sm, item1, item2, item3));
+
+                    case ArrayReadStateMachine.ReadResult.Done:
+                        if (sm.CurrentLength is not 3)
+                            return Error("Invalid JSON value; JSON array has too few values.");
+
+                        return Value((item1, item2, item3));
```

Comment:
nit: this could be written as a single `return`, as follows?

```c#
return sm.CurrentLength is not 3
     ? Error("Invalid JSON value; JSON array has too few values.")
     : Value((item1, item2, item3));
```

Reply:
Done with e2053e2d43481bddde649b067e77f2d65e892359